editor/src/components/canvas/canvas-strategies/strategies/grid-helpers.ts

```diff
@@ -647,3 +655,175 @@ function gridTemplateToNumbers(gridTemplate: GridTemplate | null): Array<number>
 
   return result
 }
+
+type DimensionIndexes = {
+  originalIndex: number // the index of this element in the original values
+  repeatedIndex: number // the index of this element, if it's generated via a repeat, inside the repeated values array definition
+}
+
+export type ExpandedGridDimension = GridDimension & {
+  indexes: DimensionIndexes
+}
+
+function expandedGridDimension(
+  dim: GridDimension,
+  originalIndex: number,
+  repeatedIndex: number = 0,
+): ExpandedGridDimension {
+  return {
+    ...dim,
+    indexes: {
+      originalIndex: originalIndex,
+      repeatedIndex: repeatedIndex,
+    },
+  }
+}
+
+export function expandGridDimensions(template: GridDimension[]): ExpandedGridDimension[] {
+  // Expanded representation of the original values, where repeated elements are serialized.
+  // Each element also contains the indexes information to be used later on to build the resized
+  // template string.
+  return template.reduce((acc, cur, index) => {
+    if (isGridCSSRepeat(cur)) {
+      const repeatGroup = cur.value.map((dim, repeatedIndex) =>
+        expandedGridDimension(dim, index, repeatedIndex),
+      )
+      let expanded: ExpandedGridDimension[] = []
+      for (let i = 0; i < cur.times; i++) {
+        expanded.push(...repeatGroup)
+      }
+      return [...acc, ...expanded]
+    } else {
+      return [...acc, expandedGridDimension(cur, index)]
+    }
+  }, [] as ExpandedGridDimension[])
+}
+
+function alterGridDimensions(params: {
+  originalValues: GridDimension[]
+  target: ExpandedGridDimension
+  action: AlterGridTemplateDimensionAction
+}): GridDimension[] {
+  return mapDropNulls((dim, index) => {
+    if (index !== params.target.indexes.originalIndex) {
+      return dim
+    } else if (isGridCSSRepeat(dim)) {
+      const repeatedIndex = params.target.indexes.repeatedIndex ?? 0
+      const before = dim.value.slice(0, repeatedIndex)
+      const after = dim.value.slice(repeatedIndex + 1)
+      switch (params.action.type) {
+        case 'REMOVE':
+          if (before.length + after.length === 0) {
+            return null
+          }
+          return gridCSSRepeat(dim.times, [...before, ...after])
+        case 'REPLACE':
+          return gridCSSRepeat(dim.times, [...before, params.action.newValue, ...after])
+        default:
+          assertNever(params.action)
+      }
+    } else {
+      switch (params.action.type) {
+        case 'REPLACE':
+          return params.action.newValue
+        case 'REMOVE':
+          return null
+        default:
+          assertNever(params.action)
+      }
+    }
+  }, params.originalValues)
+}
+
+export type ReplaceGridDimensionAction = {
+  type: 'REPLACE'
+  newValue: GridDimension
+}
+
+export type RemoveGridDimensionAction = {
+  type: 'REMOVE'
+}
+
+export type AlterGridTemplateDimensionAction =
```

Comment:
I don't really like the Action terminology, we use that for our dispatchable actions, someone would easily think these actions are EditorActions too.

Reply:
ah good point